nvme_gw_persistence.py

```diff
@@ -0,0 +1,404 @@
+        except rados.ObjectExists:
+            self.logger.info(f"{self.omap_name} omap object already exists.")
+
+    def _write_key(self, key: str, val: str):
+        """Writes key and value to the persistent config."""
+
+        try:
+            version_update = int(self.version) + 1
+            with rados.WriteOpCtx() as write_op:
+                # Compare operation failure will cause write failure
+                write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
+                                  rados.LIBRADOS_CMPXATTR_OP_EQ)
+                self.ioctx.set_omap(write_op, (key,), (val,))
+                self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                    (str(version_update),))
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+            self.version = version_update
+            self.logger.debug(f"omap_key generated: {key}")
+        except Exception as ex:
+            self.logger.error(f"Unable to write to omap: {ex}. Exiting!")
+            raise
+
+    def _delete_key(self, key: str):
+        """Deletes key from omap persistent config."""
+
+        version_update = int(self.version) + 1
+        with rados.WriteOpCtx() as write_op:
+            # Compare operation failure will cause delete failure
+            write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
+                              rados.LIBRADOS_CMPXATTR_OP_EQ)
+            self.ioctx.remove_omap_keys(write_op, (key,))
+            self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                (str(version_update),))
+            self.ioctx.operate_write_op(write_op, self.omap_name)
+        self.version = version_update
+        self.logger.debug(f"omap_key deleted: {key}")
+
+    def add_bdev(self, bdev_name: str, val: str):
+        """Adds a bdev to the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._write_key(key, val)
+
+    def delete_bdev(self, bdev_name: str):
+        """Deletes a bdev from the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._delete_key(key)
+
+    def _restore_bdevs(self, omap_dict, callback):
+        """Restores a bdev from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.BDEV_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                req = pb2.bdev_create_req(
+                    bdev_name=args["bdev_name"],
+                    ceph_pool_name=args["ceph_pool_name"],
+                    rbd_name=args["rbd_name"],
+                    block_size=int(args["block_size"]),
+                )
+                ret = callback(req)
+
+    def add_namespace(self, subsystem_nqn: str, bdev_name: str, val: str):
+        """Adds a namespace to the persistent config."""
+        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + bdev_name
```

Comment:
AFAIR there is no restriction to use '_' in nqn and bdev. So one can have:
```
subsystem_nqn = 'a'
bdev_name = 'b_c'
```
and 
```
subsystem_nqn = 'a_b'
bdev_name = 'c'
```
And it will produce the same keys. I suppose we need a special separator that is forbidden at least in bdev name (even if tgt does not have this restriction we can add it to the gateway).

Note, I have not looked carefully on other cases where we build keys with '_' separator. We might have similar problems there, e.g. for listener keys.

Reply:
For now I've changed the namespace key to be `self.NAMEPSACE_PREFIX + subsystem_nqn + "_" + nsid`. NSIDs cannot contain underscores and are unique within the subsystem. This also helps ensure namespaces using the same id on restore. 

This is certainly still a problem for host keys. I created an issue at #10 to keep track of this.